### ISSUE TITLE
cant find the io-dependency

### ISSUE THREAD

**Issue by foreversun69** [TotallyInformation/node-red-contrib-uibuilder#194]
Hi,

my problem, i can't find the right path for the io-dependency. I have try the example from https://github.com/TotallyInformation/node-red-contrib-uibuilder/wiki/Example:-ReactJS without and with project, but both approches have the same problem with the io-dependency.

Software       | Version
-------------- | -------
Node.JS        |  v16.14.0
npm            |  8.15.1
Node-RED       | v3.0.1
uibuilder node | 5.1.1
uibuilderFE    | 5.1.1
OS             | Ubuntu 20.04
Browser        | Firefox 108.0

### How is Node-RED installed? Where is uibuilder installed?

16 Dec 07:48:02 - [info] Settings file  : ~/.node-red/settings.js
16 Dec 07:48:02 - [info] Context store  : 'default' [module=memory]
16 Dec 07:48:02 - [info] User directory : ~/.node-red
16 Dec 07:48:02 - [info] Projects directory: ~/.node-red/projects
16 Dec 07:48:02 - [info] Server now running at http://127.0.0.1:1880/
16 Dec 07:48:02 - [info] Active project : testproject
16 Dec 07:48:02 - [info] Flows file     : ~/.node-red/projects/testproject/flows.json
16 Dec 07:48:02 - [info] +-----------------------------------------------------
16 Dec 07:48:02 - [info] | uibuilder v5.1.1 initialised
16 Dec 07:48:02 - [info] | root folder: ~/.node-red/projects/testproject/uibuilder
16 Dec 07:48:02 - [info] | Using Node-RED's webserver at:
16 Dec 07:48:02 - [info] |   http://0.0.0.0:1880/
16 Dec 07:48:02 - [info] | Installed packages:
16 Dec 07:48:02 - [info] |   react
16 Dec 07:48:02 - [info] +-----------------------------------------------------

![image](https://user-images.githubusercontent.com/44213573/208104151-fe4842f8-1b46-4294-8b19-5cbc0445bbf8.png)

```js
uibuilder.start({
            namespace: "http://127.0.0.1:1880/react_ui", 
            ioPath: "http://127.0.0.1:1880/uibuilder/vendor/socket.io"
        });
```

can you please help me?

cheers Patrick

**Comment by TotallyInformation** [comment]
I assume you tried without specifying the ioPath - uibuilder v5+ should normally always be able to find it itself? Also, have you changed httpNodeRoot in settings.js - if so, that impacts the paths.

Also note that you don't need to specify a full URL for namespace unless you are serving your front-end code from a different web server than node-red's. See https://totallyinformation.github.io/node-red-contrib-uibuilder/#/front-end-library?id=parameters for details.

**Comment by foreversun69** [comment]
Hi,

- per default ( `uibuilder.start()` without set any parameter ), the ioPath is: `/uibuilder/vendor/socket.io`.
- httpNodeRoot in settings.js is default.
- The picture shows the default-values:
![grafik](https://user-images.githubusercontent.com/44213573/208631157-60263a03-4a35-4b26-b8dc-2aecf1f1a04e.png)

the front-end runs on port 3000 in a different web server.

cheers Patrick

**Comment by TotallyInformation** [comment]
Are you still using the old uibuilderfe client? If so, I'm afraid you do need to still have the reference to the socket.io client in your HTML. Only the new clients (recommended) don't need it.

for the new clients, the html is:

```html
    <!-- #region Supporting Scripts. These MUST be in the right order. Note no leading / - socket.io no longer needed  -->
    <script src="./uibuilder.iife.min.js">/* REQUIRED: remove 'min.' to use dev version */</script>
    <script src="./index.js">/* OPTIONAL: Put your custom code in that */</script>
    <!-- #endregion -->
```

For the old client:
```html
    <!-- #region Supporting Scripts. These MUST be in the right order. Note no leading / -->
    <script src="../uibuilder/vendor/socket.io/socket.io.js">/* REQUIRED: Socket.IO is loaded only once for all instances. Without this, you don't get a websocket connection */</script>
    <script src="./uibuilderfe.min.js">/* REQUIRED: remove 'min.' to use dev version */</script>
    <script src="./index.js">/* OPTIONAL: Put your custom code here */</script>
    <!-- #endregion -->
```

Plus whatever other libraries you need of course.

I'm just re-doing all of the templates and examples for a post v6.0 release.

**Comment by foreversun69** [comment]
Hi,

**i test on the same webserver:**
- now for the old client-settings:
![grafik](https://user-images.githubusercontent.com/44213573/208737854-a935cb49-d066-43d9-acac-a49f56253546.png)

```
<script src="../uibuilder/vendor/socket.io/socket.io.js">/* REQUIRED: Socket.IO is loaded only once for all instances. Without this, you don't get a websocket connection */</script>
<script src="./uibuilderfe.min.js">/* REQUIRED: remove 'min.' to use dev version */</script>

uibuilder.start();
```

- and for the new client-settings:
![grafik](https://user-images.githubusercontent.com/44213573/208735697-1921e264-1bc3-4669-a62c-2ba0ab7874ce.png)

**for my extern server:**
![grafik](https://user-images.githubusercontent.com/44213573/208738873-131c25fd-a6e8-4b67-97fe-03819438f5ff.png)
```
<script src="http://127.0.0.1:1880/uibuilder/vendor/socket.io/socket.io.js">/* REQUIRED: Socket.IO is loaded only once for all instances. Without this, you don't get a websocket connection */</script>
<script src="http://127.0.0.1:1880/uibuilderfe.min.js">/* REQUIRED: remove 'min.' to use dev version */</script>

uibuilder.start({
      namespace: "http://127.0.0.1:1880/react_ui",
      /* ioPath: "http://127.0.0.1:1880/uibuilder/vendor/socket.io/socket.io.js", */
      vueApp: null
  });
```

thank you very much!

**Comment by foreversun69** [comment]
> Are you still using the old uibuilderfe client? If so, I'm afraid you do need to still have the reference to the socket.io client in your HTML. Only the new clients (recommended) don't need it.

i use `import uibuilder from 'node-red-contrib-uibuilder/front-end/uibuilderfe';`. I think, this is the old client, where is the new client?

**Comment by TotallyInformation** [comment]
You are mixing old and new. If you want to do module imports in your front-end code, you should be using the ESM version of the new library. `uibuilder.esm.js` or `uibuilder.esm.min.js`.

You have imported the old version which would mean that you would need to also import the socket.io client library. However, the old code was not designed to work with ES Modules and I'm not sure what other side-effects there might be. The v5 releases introduced the new IIFE and ESM libraries which are all based off a single codebase which is a lot more modern and a lot more robust. Now that v6 is out, the new libraries are now the default and the old library can quietly slink off into the sunset 😼

**Comment by foreversun69** [comment]
Hi,

now i have a new Problem
![grafik](https://user-images.githubusercontent.com/44213573/208777391-5e4570fc-5d15-4b30-a0c9-94d3f6d898a9.png)

my setup:
```
import uibuilder from 'node-red-contrib-uibuilder/front-end/uibuilder.esm';

<script src="./uibuilder.iife.min.js">/* REQUIRED: remove 'min.' to use dev version */</script>

uibuilder.start();
```

**Comment by TotallyInformation** [comment]
Sorry, I think we have some crossed wires. Let's back up a bit. So you are using a build step to build a REACT UI?

Generally, I recommend that you don't include uibuilder in your build step but rather exclude it. Then you don't need to `import` it. Just use the link statement in your HTML.

Since you are using Babel and REACT, you shouldn't in fact use the ESM version since that is, as the name implies, for ES Modules, you need the more traditional IIFE version.

**Comment by foreversun69** [comment]
Hi,

oh sorry for that missunderstanding spelling: Yes i use all the time the https://github.com/TotallyInformation/node-red-contrib-uibuilder/wiki/Example:-ReactJS **_React-Example_**. Usually i dont bind any <script>-Tags in the index.html. All the dependency i load via `import` direct in the components.

I have tried out the iife and esm ( you have wrote in the last text ) version, but no of them will work. The only way, that will work for me is the mixed old-client version:
```
import uibuilder from 'node-red-contrib-uibuilder/front-end/uibuilderfe';

<!-- old Client extern  -->
<script src="http://127.0.0.1:1880/uibuilder/vendor/socket.io/socket.io.js">/* REQUIRED: Socket.IO is loaded only once for all instances. Without this, you don't get a websocket connection */</script>
<script src="http://127.0.0.1:1880/uibuilderfe.min.js">/* REQUIRED: remove 'min.' to use dev version */</script>

<!-- old Client intern  -->
<!-- <script src="../uibuilder/vendor/socket.io/socket.io.js">/* REQUIRED: Socket.IO is loaded only once for all instances. Without this, you don't get a websocket connection */</script>
<script src="./uibuilderfe.min.js">/* REQUIRED: remove 'min.' to use dev version */</script> -->
```
i thought the mixed version as an experimental environment or playground is ok. React is my "Main-Framework", alternative i can use the svelte or plain-version from uibuilder ( they will work without any "problems" ).

**Comment by TotallyInformation** [comment]
OK, I think I've probably confused things rather than helping. I don't use REACT so I'm not really familiar with the tools. Let me follow the WIKI example and see if I can get things working.

A "build step" is sometimes (wrongly) referred to as a "compile". I takes your source code and processes it to a built version ready to run. Along the way, it may do things like pull multiple files into a single js file and convert from newer to older versions of JavaScript.

One of the confusions here is about the `import` statement. Natively, browser JavaScript only supports this for ES Modules which only work in modern browsers. Your build step however allows `import` and works out how to combine the external files into your output file so that it works with older browsers. I do something similar with the new uibuilder libraries where I use a build tool called ESBUILD which, among other things, locks in the socket.io client.

Bear with me while I do some testing.

**Comment by TotallyInformation** [comment]
As a taster though, here is an example that I'd forgotten I had - I'll create a WIKI from it. It uses REACT but does not need a build step.

index.html
```
<!doctype html>
<html lang="en"><head>

    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <title>Node-RED UI Builder - REACTjs template</title>
    <meta name="description" content="Node-RED UI Builder - REACTjs template">
    <link rel="icon" href="./images/node-blue.ico">

    <link type="text/css" rel="stylesheet" href="./index.css" media="all">

</head><body>
    
    <h1>uibuilder REACTjs Template</h1>
    <p>Based on React in One Minute</p>
    <p>This page demonstrates using React with no build tooling.</p>
    <p>React is loaded as a script tag.</p>

    <!-- We will put our React component inside this div. -->
    <div id="like_button_container"></div>

    <pre id="msg" class="syntax-highlight">Waiting for a message from Node-RED</pre>

    <!-- You need socket.io if using the old client but the new client works fine on its own -->
    <!-- <script src="../uibuilder/vendor/socket.io/socket.io.js"></script> -->
    <!-- <script src="./uibuilderfe.min.js"></script> -->
    <script src="./uibuilder.iife.min.js"></script>
    
    <!-- Load React. -->
    <!-- Note: when deploying, replace "development.js" with "production.min.js". -->
    <script src="https://unpkg.com/react@17/umd/react.development.js" crossorigin></script>
    <script src="https://unpkg.com/react-dom@17/umd/react-dom.development.js" crossorigin></script>

    <script src="./index.js"></script>

</body></html>
```

index.js
```
/* jshint browser: true, esversion: 6, asi: true */
/* globals uibuilder */
// @ts-nocheck
'use strict'

/** Minimalist code for uibuilder and Node-RED */

// return formatted HTML version of JSON object
function syntaxHighlight(json) {
    json = JSON.stringify(json, undefined, 4)
    json = json.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;')
    json = json.replace(/("(\\u[a-zA-Z0-9]{4}|\\[^u]|[^\\"])*"(\s*:)?|\b(true|false|null)\b|-?\d+(?:\.\d*)?(?:[eE][+\-]?\d+)?)/g, function (match) {
        var cls = 'number'
        if (/^"/.test(match)) {
            if (/:$/.test(match)) {
                cls = 'key'
            } else {
                cls = 'string'
            }
        } else if (/true|false/.test(match)) {
            cls = 'boolean'
        } else if (/null/.test(match)) {
            cls = 'null'
        }
        return '<span class="' + cls + '">' + match + '</span>'
    })
    return json
} // --- End of syntaxHighlight --- //

const e = React.createElement

class LikeButton extends React.Component {
  constructor(props) {
    super(props)
    this.state = { liked: false }
  }

  render() {
    if (this.state.liked) {
      return 'You liked this.'
    }

    return e(
      'button',
      { onClick: () => {
          this.setState({ liked: true }) 
          uibuilder.send({payload: {liked: true}})
      }},
      'Like'
    );
  }
}

const domContainer = document.querySelector('#like_button_container')
ReactDOM.render(e(LikeButton), domContainer)


// run this function when the document is loaded
window.onload = function() {
    // Start up uibuilder - not needed with the new client
    // uibuilder.start()

    // Listen for incoming messages from Node-RED
    uibuilder.onChange('msg', function(msg){
        console.info('[indexjs:uibuilder.onChange] msg received from Node-RED server:', msg)

        // dump the msg as text to the "msg" html element
        const eMsg = document.getElementById('msg')
        eMsg.innerHTML = syntaxHighlight(msg)
    })
}
```

**Comment by TotallyInformation** [comment]
Well, that was "fun", not!

I've added a new "issue" to @gaillarddamien's repo explaining what the issues are and how to fix them.

https://github.com/gaillarddamien/uibuilder-react-example/issues/2

Could you please check through that post and see if you can get it working? I got it working following the instructions I've given with the exception that the dev server doesn't work. But the build does so you can access the resulting page and comms to/from Node-RED work correctly.

**Comment by foreversun69** [comment]
Hi,

in my case i test the old fashion way includes the improves they you have written in the description, after `npm start` or `npm run build` following error:
![grafik](https://user-images.githubusercontent.com/44213573/209002602-a14c602c-5af4-4e0f-9cfc-f07d5d5c922e.png)

**Comment by TotallyInformation** [comment]
Now you are hitting another webpack issue rather than a uibuilder one.  Try deleting the `.eslintrc.js` file - that should deal with that issue.

**Comment by foreversun69** [comment]
yes it deals! But now:
![grafik](https://user-images.githubusercontent.com/44213573/209004278-245608b6-e362-4539-9bbf-97b211732ff7.png)

**Comment by TotallyInformation** [comment]
Can you please share your full package.json, index.html and index.js files.

**Comment by foreversun69** [comment]
```json
{
  "name": "uibuilder-react-example",
  "version": "0.1.0",
  "private": true,
  "devDependencies": {
    "@testing-library/jest-dom": "^5.13.0",
    "@testing-library/react": "^11.2.7",
    "@testing-library/user-event": "^12.8.3",
    "node-red-contrib-uibuilder": "*",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-scripts": "4.0.3",
    "socket.io-client": "^4.5.4",
    "web-vitals": "^1.1.2"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject"
  },
  "eslintConfig": {
    "extends": [
      "react-app",
      "react-app/jest"
    ]
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "proxy": "http://localhost:1880"
}
```
```html
<!DOCTYPE html>
<html lang="en"><head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="theme-color" content="#000000">

    <link rel="manifest" href="%PUBLIC_URL%/manifest.json">
    <link rel="shortcut icon" href="%PUBLIC_URL%/favicon.ico">

    <title>React UI - uibuilder and Node-RED</title>
</head><body>

    <div>
        <h1>
            Welcome to REACT-UIbuilder for Node-RED!
        </h1>
    </div>
    <div id="root">
        <div id="app">
        </div>
    </div>

</body>
</html>
```
```js
import React from 'react';
import ReactDOM from 'react-dom';
import './index.css';
import App from './App';
import reportWebVitals from './reportWebVitals';

ReactDOM.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>,
  document.getElementById('root')
);

// If you want to start measuring performance in your app, pass a function
// to log results (for example: reportWebVitals(console.log))
// or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
reportWebVitals();
```

**Comment by TotallyInformation** [comment]
Sorry, I edited the post to ask for `/src/UserData.js`

**Comment by foreversun69** [comment]
ok i also edit.

**Comment by foreversun69** [comment]
This is the running version:
```json
{
  "name": "react_ui",
  "version": "0.1.0",
  "private": true,
  "proxy": "http://localhost:1880/react_ui",
  "dependencies": {
    "@testing-library/jest-dom": "^5.16.5",
    "@testing-library/react": "^13.4.0",
    "@testing-library/user-event": "^13.5.0",
    "node-red-contrib-uibuilder": "file:../../../../node_modules/node-red-contrib-uibuilder",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-scripts": "5.0.1",
    "web-vitals": "^2.1.4"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject"
  },
  "eslintConfig": {
    "extends": [
      "react-app",
      "react-app/jest"
    ]
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  }
}
```

```html
<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="utf-8" />
  <link rel="icon" href="%PUBLIC_URL%/favicon.ico" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="theme-color" content="#000000" />
  <meta name="description" content="Web site created using create-react-app" />
  <link rel="apple-touch-icon" href="%PUBLIC_URL%/logo192.png" />
  <!--
      manifest.json provides metadata used when your web app is installed on a
      user's mobile device or desktop. See https://developers.google.com/web/fundamentals/web-app-manifest/
    -->
  <link rel="manifest" href="%PUBLIC_URL%/manifest.json" />
  <!--
      Notice the use of %PUBLIC_URL% in the tags above.
      It will be replaced with the URL of the `public` folder during the build.
      Only files inside the `public` folder can be referenced from the HTML.

      Unlike "/favicon.ico" or "favicon.ico", "%PUBLIC_URL%/favicon.ico" will
      work correctly both with client-side routing and a non-root public URL.
      Learn how to configure a non-root public URL by running `npm run build`.
    -->
  <title>React App</title>
</head>

<body>
  <noscript>You need to enable JavaScript to run this app.</noscript>
  <div id="root"></div>
  <!--
      This HTML file is a template.
      If you open it directly in the browser, you will see an empty page.

      You can add webfonts, meta tags, or analytics to this file.
      The build step will place the bundled scripts into the <body> tag.

      To begin the development, run `npm start` or `yarn start`.
      To create a production bundle, use `npm run build` or `yarn build`.
    -->
</body>
<!-- <script src="uibuilder.iife.min.js">/* REQUIRED: remove 'min.' to use dev version */</script> -->
<!-- <script type="module" src="uibuilder.esm.min.js">/* REQUIRED: remove 'min.' to use dev version */</script> -->

<!-- old Client extern  -->
<script src="http://127.0.0.1:1880/uibuilder/vendor/socket.io/socket.io.js">/* REQUIRED: Socket.IO is loaded only once for all instances. Without this, you don't get a websocket connection */</script>
<script src="http://127.0.0.1:1880/uibuilderfe.min.js">/* REQUIRED: remove 'min.' to use dev version */</script>

<!-- old Client intern  -->
<!-- <script src="../uibuilder/vendor/socket.io/socket.io.js">/* REQUIRED: Socket.IO is loaded only once for all instances. Without this, you don't get a websocket connection */</script>
<script src="./uibuilderfe.min.js">/* REQUIRED: remove 'min.' to use dev version */</script> -->

</html>
```

```js
/* This is where the uibuilder action happens */
import React, { Component, useState, useEffect } from 'react';
/* import ReactDOM from 'react-dom';
import { findDOMNode } from 'react-dom'; */

import uibuilder from 'node-red-contrib-uibuilder/front-end/uibuilderfe';

function UserData(props) {

    // Example of retrieving data from uibuilder
    const [feVersion, setFeVersion] = useState(uibuilder.get('version'));
    const [socketConnectedState, setSocketConnectedState] = useState(false);
    const [serverTimeOffset, setServerTimeOffset] = useState('[unknown]');
    const [msgRecvd, setMsgRecvd] = useState('[Nothing]');
    const [msgsReceived, setMsgsReceived] = useState(0);
    const [msgCtrl, setMsgCtrl] = useState('[Nothing]');
    const [msgsControl, setMsgsControl] = useState(0);
    const [msgSent, setMsgSent] = useState('[Nothing]');
    const [msgsSent, setMsgsSent] = useState(0);
    const [msgCtrlSent, setMsgCtrlSent] = useState('[Nothing]');
    const [msgsCtrlSent, setMsgsCtrlSent] = useState(0);

    const onHandleClick = (event) => {
        console.log(event);
        uibuilder.send("HELLO");

    };

    useEffect(
        () => {
            /** **REQUIRED** Start uibuilder comms with Node-RED @since v2.0.0-dev3
             * Pass the namespace and ioPath variables if hosting page is not in the instance root folder
             * e.g. If you get continual `uibuilderfe:ioSetup: SOCKET CONNECT ERROR` error messages.
             * e.g. uibuilder.start('/nr/uib', '/nr/uibuilder/vendor/socket.io') // change to use your paths/names
             */
            /* uibuilder.start(); */
            uibuilder.start({
                namespace: "http://127.0.0.1:1880/react_ui",
                //ioPath: "http://127.0.0.1:1880/uibuilder/vendor/socket.io/socket.io.js",
                // vueApp: null
            });

            /** You can use the following to help trace how messages flow back and forth.
             * You can then amend this processing to suite your requirements.
             */

            //#region ---- Trace Received Messages ---- //
            // If msg changes - msg is updated when a standard msg is received from Node-RED over Socket.IO
            // newVal relates to the attribute being listened to.
            uibuilder.onChange('msg', (newVal) => {
                console.info('[uibuilder.onChange] msg received from Node-RED server:', newVal);
                setMsgRecvd(newVal);
            });

            // As we receive new messages, we get an updated count as well
            uibuilder.onChange('msgsReceived', (newVal) => {
                console.info('[uibuilder.onChange] Updated count of received msgs:', newVal);
                setMsgsReceived(newVal);
            });

            // If we receive a control message from Node-RED, we can get the new data here - we pass it to a Vue variable
            uibuilder.onChange('ctrlMsg', (newVal) => {
                console.info('[uibuilder.onChange:ctrlMsg] CONTROL msg received from Node-RED server:', newVal);
                setMsgCtrl(newVal);
            });

            // Updated count of control messages received
            uibuilder.onChange('msgsCtrl', (newVal) => {
                console.info('[uibuilder.onChange:msgsCtrl] Updated count of received CONTROL msgs:', newVal);
                setMsgsControl(newVal);
            });

            //#endregion ---- End of Trace Received Messages ---- //

            //#region ---- Trace Sent Messages ---- //
            // You probably only need these to help you understand the order of processing //
            // If a message is sent back to Node-RED, we can grab a copy here if we want to
            uibuilder.onChange('sentMsg', (newVal) => {
                console.info('[uibuilder.onChange:sentMsg] msg sent to Node-RED server:', newVal);
                setMsgSent(newVal);
            });

            // Updated count of sent messages
            uibuilder.onChange('msgsSent', (newVal) => {
                console.info('[uibuilder.onChange:msgsSent] Updated count of msgs sent:', newVal);
                setMsgsSent(newVal);
            });

            // If we send a control message to Node-RED, we can get a copy of it here
            uibuilder.onChange('sentCtrlMsg', (newVal) => {
                console.info('[uibuilder.onChange:sentCtrlMsg] Control message sent to Node-RED server:', newVal);
                setMsgCtrlSent(newVal);
            });

            // And we can get an updated count
            uibuilder.onChange('msgsSentCtrl', (newVal) => {
                console.info('[uibuilder.onChange:msgsSentCtrl] Updated count of CONTROL msgs sent:', newVal);
                setMsgsCtrlSent(newVal);
            });

            //#endregion ---- End of Trace Sent Messages ---- //

            // If Socket.IO connects/disconnects, we get true/false here
            uibuilder.onChange('ioConnected', (newVal) => {
                console.info('[uibuilder.onChange:ioConnected] Socket.IO Connection Status Changed to:', newVal)
                setSocketConnectedState(newVal);
            });

            // If Server Time Offset changes
            uibuilder.onChange('serverTimeOffset', (newVal) => {
                console.info('[uibuilder.onChange:serverTimeOffset] Offset of time between the browser and the server has changed to:', newVal)
                setServerTimeOffset(newVal);
            });

            //Manually send a message back to Node-RED after 2 seconds

            /* window.setTimeout(function () {
                console.log("TEST aus setTimeout")
                console.info('Sending a message back to Node-RED-after2sdelay')
                uibuilder.send({ 'topic': 'uibuilderfe', 'payload': 'I am a message sent from the uibuilder front end' })
            }, 2000);


            return () => {
                uibuilder.clearEventListeners();
            }; */

        },
        []
    );

    return (

        <div style={{ height: "50vh" }}>
            <hr></hr>
            <div className="d1">
                <div>Last msg Received:</div>
                <pre><code>{JSON.stringify(msgRecvd, null, 2)}</code></pre>
                <div># Msgs Received: {msgsReceived}</div>
            </div>

            <div className="d1">
                <div>last Ctl Msg Received:</div>
                <pre><code>{JSON.stringify(msgCtrl, null, 2)}</code></pre>
                <div># Control Msgs Received: {msgsControl}</div>
            </div>

            <div className="d1">
                <div>last Msg Sent</div>
                <pre><code>{JSON.stringify(msgSent, null, 2)}</code></pre>
                <div># msgs Sent: {msgsSent}</div>
            </div>

            <div className="d1">
                <div>Socket Connected?: {socketConnectedState}</div>
                <div>uibuilderfe Version: {feVersion}</div>
                <div>Server Time Offset from browser: {serverTimeOffset}</div>
            </div>
            <button onClick={onHandleClick} >Send a message to node-red</button>

        </div>
    );

}

export default UserData;
```

**Comment by TotallyInformation** [comment]
Your `UserData.js` should have had both:

```javascript
import { io } from "socket.io-client";
import uibuilder from '../../node_modules/node-red-contrib-uibuilder/front-end/uibuilderfe'
```

I'm sure that was working earlier but it isn't for me now.

You will need to replace with:

```javascript
import uibuilder from '../../node_modules/node-red-contrib-uibuilder/src/front-end-module/uibuilder.module'
```
(don't load the socket.io client).

But the dev server won't work. Build works just fine though.

Something to do with the way that webpack and babel are interacting. I don't know enough about them to be able to fix I'm afraid.

**Comment by foreversun69** [comment]
Yes it works!

**Comment by TotallyInformation** [comment]
Phew. REACT is a LOT more complex than Vue which is why I don't use it. It has always seemed too much like hard work.

**Comment by foreversun69** [comment]
Thank you for your time! :-)
